modules/raft-api/src/main/java/org/apache/ignite/internal/raft/service/RaftGroupListener.java

```diff
@@ -93,4 +93,18 @@ default void onConfigurationCommitted(CommittedConfiguration config) {
     default void onBeforeApply(Command command) {
         // No-op.
     }
+
+    /**
+     * Returns {@code true} if {@link #onBeforeApply(Command)} and saving of the command to the LogManager
+     * must be executed atomically (that is, these actions cannot intermingle for different commands
+     * on the same Raft group.
+     *
+     * <p>This does not influence read commands: for them no atomicity guarantees are provided.
+     *
+     * @param command The command.
+     * @return {@code true} if {@link #onBeforeApply(Command)} and saving must be executed atomically.
+     */
+    default boolean atomicOnBeforeApplyAndWrite(Command command) {
```

Comment:
I would suggest to introduce a sub-class, that will declare the `onBeforeApply` method. This way we can always know if `onBeforeApply` is not a no-op and always execute it atomically with `write`

Reply:
But do we need to ALWAYS execute it atomically with the write? I think no

Reply:
Why not? I can't imagine a scenario when you need to implement `onBeforeApply` and not have it executed atomically. Having a separate boolean method seems to cryptic to me. Anyway, we can add it after we find out that we need it.

Reply:
ok, I think my proposed solution doesn't make sense, because we don't have separate listener instances for writes and reads (and we only need to call onBeforeApply for writes)